### DISCUSSION TITLE
No table grid with typst

### DISCUSSION THREAD

**Discussion by psads-git** [quarto-dev/quarto-cli#9286] [Q&A] [answered]
Consider the following `quarto` document:

````
---
format: typst
---

```{=typst}
#align(center,
    block(
      breakable: false,
      table(
      columns: (auto, auto, auto),
      align: horizon,
      [a], [b], [c],
      [1], [M1], [15],
      [2], [M2], [17],

      )
    )
  )
```
````

In the output, the table has no grid, while the _exact same_ code in `typst` app returns a table with a full grid. Could you please help me? Thanks!

**Comment by mcanouil** [comment]
Use `keep-typ: true` and investigate the Typst document you are generating using Quarto.

Additionally you can use Pandoc to see what Pandoc produces from your input:

```sh
quarto pandoc file.qmd --from markdown --to typst -o file.typ
```

Also, be sure to compare things that are comparable, i.e., same version.

```sh
typst --version
quarto typst --version
```

Or better, use the Typst embedded in Quarto.

**Comment by psads-git** [comment]
Thanks, @mcanouil, for your reply.

> Use `keep-typ: true` and investigate the Typst document you are generating using Quarto.

It produces the exact same `typst` code I used in the `quarto` document to create the table.

> Additionally you can use Pandoc to see what Pandoc produces from your input:
> 
> ```shell
> quarto pandoc file.qmd --from markdown --to typst -o file.typ
> ```

It produces the exact same `typst` code I used in the `quarto` document to create the table.

> Also, be sure to compare things that are comparable, i.e., same version.
> 
> ```shell
> typst --version
> quarto typst --version
> ```

```
$ typst --version
quarto typst --version
typst 0.11.0 (unknown hash)
typst 0.10.0 (70ca0d25)
$
``` 

> Or better, use the Typst embedded in Quarto.

It returns the table with full grid.

In short, the culprit is most probably the `typst` code of the `quarto` template.

**Reply by mcanouil** [reply]
You are not using the same version of Typst.
Typst 0.11 brings a lot of stuff for tables and is not avalaible as showed by your output in Quarto 1.4.

The issue is not with Quarto since you said yourself that the output is the same with Pandoc.
The template and support for Typst comes from Pandoc.

Typst 0.11 is or will be in 1.5.

Expecting two versions of Typst to behave the same way is not a realistic expectation I'm afraid.m, especially if you take a look at Typst changelog.

**Comment by gordonwoodhull** [comment, marked as answer]
Yes, Pandoc changed the Typst template for this case, and although we have not pulled in the latest Pandoc yet for 1.5, we pulled in this part of the template. 

The style, with lines only below the header and above the footer, is standard across formats for Pandoc, and this is why it changed. 

Here is the commit: 8f6efd2

If you want to add grid lines back, you can add a `{=typst}` raw block, something like

````
```{=typst}
#set table(
  stroke: 1pt
)
```
````
It's also worth checking the changelog to see if a change is intentional. 

https://github.com/quarto-dev/quarto-cli/blob/b24caa36965240404d0bffcaca3b2235ebae45a3/news/changelog-1.5.md?plain=1#L59

**Reply by mcanouil** [reply]
Indeed, the part setting the theme globally is:

```typst
#set table(
  inset: 6pt,
  stroke: none
)
```

Note that the "inset" part was originally hardcoded in each tables: https://github.com/jgm/pandoc/issues/9580

The above code can be found at line 239 using `keep-typ: true` and investigating the code as suggested initially.

````qmd
---
format: typst
keep-typ: true
---

```{=typst}
#align(center,
  block(
    breakable: false,
    table(
      columns: (auto, auto, auto),
      align: horizon,
      [a], [b], [c],
      [1], [M1], [15],
      [2], [M2], [17],
    )
  )
)
```
````

Then you can easily set globally whatever preference you want:

````qmd
---
format: typst
---

```{=typst}
#set table(
  stroke: 1pt
)
```

```{=typst}
#align(center,
  block(
    breakable: false,
    table(
      columns: (auto, auto, auto),
      align: horizon,
      [a], [b], [c],
      [1], [M1], [15],
      [2], [M2], [17],
    )
  )
)
```
````

Edit: and of course, you can set whatever you want in the table itself:

````qmd
---
format: typst
---

```{=typst}
#align(center,
  block(
    breakable: false,
    table(
      columns: (auto, auto, auto),
      align: horizon,
      stroke: 1pt,
      [a], [b], [c],
      [1], [M1], [15],
      [2], [M2], [17],
    )
  )
)
```
````

As a reminder for any format supported by Quarto:

1. Use `keep-*` option to look at the document produced by Quarto/Pandoc before the third party (e.g., LaTeX engines, Typst, etc.) compiles it. (for HTML, use your favourite browser developer/inspect mode)
2. Learn about the third party tools and the language. Typst has a great documentation that you really should consider to bookmark it in your favourite browser.
   You need to learn Typst in order to make non default things with it. There are no shortcuts.
4. Use Quarto partials/template, `include-` YAML keys or raw block code to integrate your code.

The process is always the same.

**Reply by psads-git** [reply]
Thanks, @mcanouil and @gordonwoodhull, for your answers.
